src/main/scala/mesosphere/marathon/core/deployment/ScalingProposition.scala

```diff
@@ -2,50 +2,51 @@ package mesosphere.marathon
 package core.deployment
 
 import mesosphere.marathon.core.condition.Condition
-import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.core.condition.Condition.UnreachableInactive
+import mesosphere.marathon.core.instance.{Goal, Instance}
 import mesosphere.marathon.state.{KillSelection, Timestamp}
 
 case class ScalingProposition(tasksToKill: Option[Seq[Instance]], tasksToStart: Option[Int])
 
 object ScalingProposition {
 
   def propose(
-    runningTasks: Seq[Instance],
-    toKill: Option[Seq[Instance]],
+    instances: Seq[Instance],
+    toDecommission: Option[Seq[Instance]],
     meetConstraints: ((Seq[Instance], Int) => Seq[Instance]),
     scaleTo: Int,
     killSelection: KillSelection): ScalingProposition = {
 
-    val killingTaskCount = runningTasks.count(_.state.condition == Condition.Killing)
+    val instancesGoalRunning: Map[Instance.Id, Instance] = instances
+      .filter(_.state.goal == Goal.Running)
+      .map(instance => instance.instanceId -> instance)(collection.breakOut)
+    val toDecommissionMap: Map[Instance.Id, Instance] = toDecommission.getOrElse(Seq.empty).map(instance => instance.instanceId -> instance)(collection.breakOut)
 
-    val runningTaskMap = Instance.instancesById(runningTasks)
-    val toKillMap = Instance.instancesById(toKill.getOrElse(Seq.empty))
-
-    val (sentencedAndRunningMap, notSentencedAndRunningMap) = runningTaskMap partition {
-      case (k, v) =>
-        toKillMap.contains(k)
+    val (sentencedAndRunningMap, notSentencedAndRunningMap) = instancesGoalRunning partition {
+      case (instanceId, instance) =>
+        toDecommissionMap.contains(instanceId) || instance.state.condition == UnreachableInactive
     }
     // overall number of tasks that need to be killed
-    val killCount = math.max(runningTasks.size - killingTaskCount - scaleTo, sentencedAndRunningMap.size)
+    val decommissionCount = math.max(instances.size - scaleTo, sentencedAndRunningMap.size)
     // tasks that should be killed to meet constraints – pass notSentenced & consider the sentenced 'already killed'
     val killToMeetConstraints = meetConstraints(
       notSentencedAndRunningMap.values.to[Seq],
-      killCount - sentencedAndRunningMap.size
+      decommissionCount - sentencedAndRunningMap.size
     )
 
     // rest are tasks that are not sentenced and need not be killed to meet constraints
     val rest: Seq[Instance] = (notSentencedAndRunningMap -- killToMeetConstraints.map(_.instanceId)).values.to[Seq]
 
-    val ordered =
+    val orderedDecommissionCandidates =
       Seq(sentencedAndRunningMap.values, killToMeetConstraints, rest.sortWith(sortByConditionAndDate(killSelection))).flatten
 
-    val candidatesToKill = ordered.take(killCount)
-    val numberOfTasksToStart = scaleTo - runningTasks.size + killCount
+    val candidatesToDecommission = orderedDecommissionCandidates.take(decommissionCount)
+    val numberOfInstancesToStart = scaleTo - instances.size + decommissionCount
```

Comment:
Isn't this wrong? `instances` are not the instances with goal running.

Reply:
Thanks!